### PR TITLE
Support submitit setup parameter

### PR DESCRIPTION
**Pull request by jgbos**
## Motivation

Add support to Submitit Slurm parameter `setup`.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/facebookresearch/hydra/blob/master/CONTRIBUTING.md)?

Yes

## Test Plan

All plugin tests pass.

## Related Issues and PRs

See #1227

### CODE REVIEW

File: plugins/hydra_submitit_launcher/hydra_plugins/hydra_submitit_launcher/submitit_launcher.py [hunk 1]
```diff
@@ -119,6 +119,13 @@ def launch(
             for x, y in params.items()
             if x not in init_keys
         }
+        # submitit `setup` parameter must be a primitive list container
+        if f"{self._EXECUTOR}_setup" in params and isinstance(
+            params[f"{self._EXECUTOR}_setup"], ListConfig
+        ):
+            params[f"{self._EXECUTOR}_setup"] = OmegaConf.to_container(
```
Comment (jieru-hu) [inline plugins/hydra_submitit_launcher/hydra_plugins/hydra_submitit_launcher/submitit_launcher.py:126]: what happens when setup = None?

also why do you need to check if its ListConfig? is it to get past mypy?

Comment (jgbos) [reply plugins/hydra_submitit_launcher/hydra_plugins/hydra_submitit_launcher/submitit_launcher.py:126]: Won't the if statement be false if it's None?  

I've been reading through the code trying to get a better handle of the assumptions for initializing the Launcher.  The code doesn't indicate that the input for initialization comes from configuration objects, can I assume this? If so, in the `__init__` function could I just do:

```python
self.params["setup"] = OmegaConf.to_container(self.params["setup"], resolve=True)
```

Comment (omry) [reply plugins/hydra_submitit_launcher/hydra_plugins/hydra_submitit_launcher/submitit_launcher.py:126]: A simpler approach is to convert all the params to primitive containers.
This seems to work:
```diff
diff --git a/plugins/hydra_submitit_launcher/hydra_plugins/hydra_submitit_launcher/submitit_launcher.py b/plugins/hydra_submitit_launcher/hydra_plugins/hydra_submitit_launcher/submitit_launcher.py
index 1b3635c7e..b862ab5b1 100644
--- a/plugins/hydra_submitit_launcher/hydra_plugins/hydra_submitit_launcher/submitit_launcher.py
+++ b/plugins/hydra_submitit_launcher/hydra_plugins/hydra_submitit_launcher/submitit_launcher.py
@@ -114,12 +114,17 @@ class BaseSubmititLauncher(Launcher):
 
         # specify resources/parameters
         baseparams = set(OmegaConf.structured(BaseQueueConf).keys())
-        params = {
-            x if x in baseparams else f"{self._EXECUTOR}_{x}": y
-            for x, y in params.items()
-            if x not in init_keys
-        }
-        executor.update_parameters(**params)
+
+        params = OmegaConf.create(
+            {
+                x if x in baseparams else f"{self._EXECUTOR}_{x}": y
+                for x, y in params.items()
+                if x not in init_keys
+            }
+        )
+
+        plain_params = OmegaConf.to_container(params, resolve=True)
+        executor.update_parameters(**plain_params)
 
         log.info(

```

You can also convert them inside the existing overly fancy dict comprehension loop (slightly more efficient but much more convoluted, at least in the current form of things.

Comment (omry) [reply plugins/hydra_submitit_launcher/hydra_plugins/hydra_submitit_launcher/submitit_launcher.py:126]: You can also just convert all the OmegaConf containers in params to containers.

**Comment by omry** [comment]
Please explain how you are testing this.

**Comment by jgbos** [comment]
I ran the tests in the folder and I also run a local (no setup) and cluster job (with setup) on our system

**Comment by jgbos** [comment]
So now I see that the plugin never "checks the schema".

**Comment by jgbos** [comment]
What do you think of adding these line to each of the Launcher classes `__init__` method (similar to the RQ Launcher code):

For `BaseSubmititLauncher`

```python
self.params = OmegaConf.structured(BaseQueueConf(**params))
````

For `LocalSubmititLauncher`

```python
self.params = OmegaConf.structured(LocalQueueConf(**params))
````

For `SlurmSubmititLauncher`

```python
self.params = OmegaConf.structured(SlurmQueueConf(**params))
self.params.setup = OmegaConf.to_container(self.params.setup, resolve=True)
````

**Comment by omry** [comment]
> So now I see that the plugin never "checks the schema".
Explain.

did you see my comment about additional_parameters not being checked?

Please show the output of `--cfg all -p hydra.launcher` when you are overriding setup before this fix attempt.

**Comment by omry** [comment]
1. You still have CI failures.
2. I was thinking about simplifying the code by converting everything passed to submitit to a primitive container.

**Comment by jgbos** [comment]
whoops, sorry.  Forgot that pushing to my fork updates this automatically.  Doing some internal tests across different platforms.  Obviously a novice at this.

1. I guess I need to dig into why ray tests are failing
2. My goal was to convert any config object to primitive containers with this line:

https://github.com/jgbos/hydra/blob/submitit-launcher-setup/plugins/hydra_submitit_launcher/hydra_plugins/hydra_submitit_launcher/submitit_launcher.py#L122

Not a big fan of the if statements there, not very clear what's going on.

**Comment by omry** [comment]
> I guess I need to dig into why ray tests are failing

Ray failures are unrelated and should be resolved soon. I missed this being only Ray that is failing.

**Comment by omry** [comment]
see #1269. try to make it work with it.

In Hydra 1.1, we should be able to use `_convert_: ALL` as an alternative. looks like here is some unrelated issue with it in plugins at the moment.

**Comment by jgbos** [comment]
Ah yes, thanks! #1269 was what I was thinking awhile ago but I the wasn't sure what I could assume for the input given it's typed with `Any`, I also didn't know about `OmegaConf.is_config` at the time.

**Review by omry — COMMENTED** [review]
Cool.
Can you add a news fragment in the news directory in the plugin?

See https://hydra.cc/docs/development/documentation#news-entries

**Comment by omry** [comment]
1. Please rebase. all tests should pass.
2. Is there a link to setup in the slurm doc we can use in the news fragment?

**Comment by jgbos** [comment]
No mention of `setup` in the docs unfortunately.  Just here in the code:

https://github.com/facebookincubator/submitit/blob/master/submitit/slurm/slurm.py#L429

**Comment by omry** [comment]
> No mention of `setup` in the docs unfortunately. Just here in the code:
> 
> https://github.com/facebookincubator/submitit/blob/master/submitit/slurm/slurm.py#L429

Is this a submitit functionality of SLURM functionality?

**Comment by jgbos** [comment]
It's for SLURM, specifically the SBATCH script.

**Comment by omry** [comment]
Not my question:
I am trying to understand if there is a documentation of a setup concept in SLURM or if this is a submitit concept.

The news file should be something like:
```
Add support for [Submitit's setup field](link to doc or code)
```

Or:
```
Add support for [SLURM sbatch setup](link to doc or code)
```

(No need to mention the conversion to primitive containers, this is not relevant to users).

**Comment by jgbos** [comment]
Sorry, yes, this is a submitit concept to allow the user to provide additional lines of code in the SBATCH script.  This would be something Submitit would need to document but has not.  So I would have to link to the code.

**Review by omry — COMMENTED** [review]
looks like this fell off the radar.
Are we ready to integrate it?

**Review by omry — APPROVED** [review]
lgtm, as far as I am concerned we can merge this.

**Comment by omry** [comment]
@shagunsodhani , would you mind releasing an updated version?
(remember to use towncrier to generate a news file).

**Comment by jgbos** [comment]
Thanks! I had assumed it was done because I've been working off the master branch, but that just means `hydra-core` and not the plugin which was just my local version 🤦.  Glad I was able to contribute and thanks for working with me on getting it right.

**Pull request merged**